internal2.lua

```diff
@@ -405,6 +413,10 @@ function Tube:walk_tube_line(pos, dir)
 	local cnt = 0
 	if dir then
 		while cnt <= self.max_tube_length do
+			local secondary = self:get_secondary_node(pos, dir)
```

Comment:
This code block irritates me. The teleport nodes are the only exception for a secondary node in a tube line. A tube line always ends at a secondary node.

Reply:
True, the only other way around that is to stop teleporters being secondary nodes that act like tubes. It is possible someone else wants to make a new block like this though.

If you're worried about how often it's executed it could be moved down a few lines until just before
425: `pos, dir = new_pos, new_dir`

That should put it after the check to see if there's another tube next, but before we move on to that tube

Reply:
`self:get_next_tube(pos, dir)` in line 420 will call `self:get_next_teleport_node(pos, dir)`. Could you do the face check there?

Reply:
Absolutely, I'll give it a shot

Reply:
I definitely think that's a better place to put it, managed to make it cleaner as well instead of using `get_secondary_node`.
We make one extra call to `get_node_lvm`, and of course one to `is_valid_dir`, but that's it.